contracts/examples/IncentivisedSlidingWindowOracle.sol

```diff
@@ -0,0 +1,176 @@
+pragma solidity =0.6.6;
+
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol';
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol';
+import '@uniswap/lib/contracts/libraries/FixedPoint.sol';
+
+import '../libraries/SafeMath.sol';
+import '../interfaces/IERC20.sol';
+import '../libraries/UniswapV2Library.sol';
+import '../libraries/UniswapV2OracleLibrary.sol';
+
+// sliding window oracle that uses observations collected over a window to provide moving price averages in the past
+// `windowSize` with a precision of `windowSize / granularity`
+// note this is a singleton oracle and only needs to be deployed once per desired parameters, which
+// differs from the simple oracle which must be deployed once per pair.
+contract IncentivisedSlidingWindowOracle {
+    using FixedPoint for *;
+    using SafeMath for uint256;
+
+    uint256 public constant ONE_HUNDRED_PERCENT = 1e18;
+
+    struct Observation {
+        uint timestamp;
+        uint price0Cumulative;
+        uint price1Cumulative;
+    }
+
+    address public immutable factory;
+    // the desired amount of time over which the moving average should be computed, e.g. 24 hours
+    uint public immutable windowSize;
+    // the number of observations stored for each pair, i.e. how many price observations are stored for the window.
+    // as granularity increases from 1, more frequent updates are needed, but moving averages become more precise.
+    // averages are computed over intervals with sizes in the range:
+    //   [windowSize - (windowSize / granularity) * 2, windowSize]
+    // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the average price for
+    //   the period:
+    //   [now - [22 hours, 24 hours (whenever it was called between these times ago)], now]
+    uint8 public immutable granularity;
+    // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+    uint public immutable periodSize;
+
+    IERC20 public immutable incentiveToken;
+    uint256 public immutable percentIncentivePerCall;
+    address public incentivisedPair;
+
+    event PriceUpdatedForPair(address pair, address tokenA, address tokenB);
+
+    // mapping from pair address to a list of price observations of that pair
+    mapping(address => Observation[]) public pairObservations;
+
+    constructor(
+        address factory_,
+        uint windowSize_,
+        uint8 granularity_,
+        IERC20 incentiveToken_,
+        uint256 percentIncentivePerCall_,
+        address incentivisedPair_
```

Comment:
Probably I'm missing more context on where we want to use an incentivized pair. I wonder if we want to be more flexible on the `incentivisedPair` we allow.